[hackathon] port addmv to OpInfo

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55545 [hackathon] port addmv to OpInfo**

Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539

Commit:
[hackathon] port addmv to OpInfo

[ghstack-poisoned]

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]

imaginary-person commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2463,6 +2486,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+               # numerical:tensor([[0.+1.0000e+00j, 0.+3.7808e+13j, 0.-9.1581e+12j, 0.+3.7517e+13j,
+               # 0.-1.5335e+13j]])
+               # analytical:tensor([[0.+1.j, 0.+1.j, 0.+1.j, 0.+1.j, 0.+1.j]])
+               SkipInfo('TestGradients', 'test_inplace_grad', dtypes=(torch.complex128,)),
```
`addmv`'s inplace variant doesn't support autograd, so `supports_inplace_autograd=False` should be used instead of skipping inplace grad tests.

Lilyjjo replied:
Thanks! I'll change it :)

imaginary-person commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2463,6 +2486,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+               # RuntimeError: "addmv_impl_cpu" not implemented for 'Half'
+               SkipInfo('TestOpInfo', 'test_supported_backward', dtypes=(torch.float16,)),
+               # RuntimeError: "addmv_impl_cpu" not implemented for 'Half'
+               SkipInfo('TestOpInfo', 'test_supported_dtypes', dtypes=(torch.float16,)),
```
I hope you wouldn't mind another unsolicited comment. :)
If `float16` isn't supported by `addmv` on CPU, then it should be removed from `dtypesIfCPU`.
BTW, you're skipping these 2 tests for CUDA as well, but the comments only mention CPU.

Lilyjjo replied:
The comments are very welcome-- I updated the tests!

imaginary-person replied:
Hello @Lilyjjo, in my previous comment above, I had only been referring to the two tests above.
It seems that you'd have to skip `test_variant_consistency_eager` and `test_out` for 'CUDA' as well, so you might want to remove the device specifier in their respective `SkipInfo`s.

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]

imaginary-person commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2463,6 +2486,36 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/55539 to track all issues for addmv
+               # failed with a large difference in tensor outputs for below tests
+               SkipInfo('TestCommon', 'test_variant_consistency_eager', dtypes=(torch.float32, torch.complex64)),
```
`test_variant_consistency_eager` might pass now, since after setting `supports_inplace_autograd=False`,  it wouldn't test the backward for its inplace variant anymore:

https://github.com/pytorch/pytorch/blob/dd2bccafc599b3931555d95c222809a94d1cbe28/test/test_ops.py#L295-L298

Lilyjjo replied:
Good idea, but I still have those tests failing even with the support_inplace_autograd=False, I'm not totally sure why. For the purpose of porting the op I don't think it's an issue

imaginary-person replied:
Thank you for your response!
I looked at your latest code, and it seems that the test is only failing on CPU, so my assumption was incorrect.
Sorry, I should've tested locally before commenting.

Lilyjjo replied:
It looks like @mruberry suggestion of changing the sample_inputs generator to check for for_inplace_variant fixed these skips 👍

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -602,6 +602,29 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
         return (input, )
 
+def sample_inputs_addmv(op_info, device, dtype, requires_grad):
+    test_cases = [((S,), (S, M), (M,), 1, 1),
```
See related issues; this has uncovered an addmv bug!

Luckily I don't think this needs to wait for those issues to be resolved, but there's a slight tweak that would be helpful here.

In https://github.com/pytorch/pytorch/pull/53014, @kshitij12345 updated test_variant_consistency_eager to pass a kwarg, "for_inplace_variant":

https://github.com/pytorch/pytorch/blob/0dff0d1537a3746e80dd3a3b655ec071fe3c2466/test/test_ops.py#L304

an example of this kwarg in use is here:

https://github.com/pytorch/pytorch/blob/0dff0d1537a3746e80dd3a3b655ec071fe3c2466/torch/testing/_internal/common_methods_invocations.py#L1810

This sample inputs can take the kwarg and do a few things to use it. One example is below:

```
for_inplace_variant = kwargs.get('for_inplace_variant', False)

test_cases = (
  ((S,), (S, M), (M,), 1, 1),
  ((S,), (S, M), (M,), 0.2, 0.6),
)

test_cases_with_broadcast = (
  ((1,), (S, M), (M,), 1, 1),
  ((1,), (S, M), (M,), 0.2, 0.6),
  ((), (S, M), (M,), 1, 1),
  ((), (S, M), (M,), 0.2, 0.6),
)

cases = test_cases if for_inplace_variant else (test_cases + test_cases_with_broadcast)
samples = []
for a_shape, b_shape, c_shape, alpha, beta in cases:
  a = make_tensor(a_shape, ...)
  b = make_tensor(b_shape, ...)
  c = make_tensor(c_shape, ...)
  sample = SampleInput(a, args=(b, c), kwargs={'alpha': alpha, 'beta': beta})
  samples.append(sample)

return tuple(samples)
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2463,6 +2486,22 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/55539 to track all issues for addmv
+               # failed with a large difference in tensor outputs for below tests
+               SkipInfo('TestCommon', 'test_variant_consistency_eager',
```
I'm curious if working around the addmv_ bug will make this skip no longer necessary

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2463,6 +2486,22 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+    OpInfo('addmv',
+           dtypes=floating_types(),
+           dtypesIfCPU=all_types_and_complex_and(torch.bfloat16),
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.complex64, torch.complex128,
```
Nice application of bfloat16 on only CUDA11OrLater

mruberry reviewed (commented):
This looks really good; thanks for participating in our hackathon porting session, @Lilyjjo! 

It definitely uncovered at least one torch.addmv_() bug, too! See linked issues. I made some inline comments for how to modify the sample inputs function in this test to handle the broadcasting. Ping me if you have any questions.

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]

Lilyjjo commented:
I updated the sample_inputs creator to handle the broadcasting! It resolved the rest of the skips except for the lack of the resizing user warning

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2523,6 +2552,19 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           skips=(
+               # issue may fix: https://github.com/pytorch/pytorch/issues/55589
+               # AssertionError: UserWarning not triggered : Resized a non-empty tensor but did not warn about it.
+               SkipInfo('TestCommon', 'test_out', dtypes=(torch.float32,), device_type='cpu'),
```
I think this needs to be skipped on CUDA, too, so if you remove `device_type='cpu'` all the tests will pass.

mruberry reviewed (approved):
Cool!

There's just one last tweak skip the out tests on CUDA (see inline comment). After that if the CI is green this is good to go!

Commit:
Update on "[hackathon] port addmv to OpInfo"


Differential Revision: [D27629053](https://our.internmc.facebook.com/intern/diff/D27629053)

Skip error tracking at: #55539


[ghstack-poisoned]
